doc/man3/SSL_CTX_set_cert_verify_callback.pod

```diff
@@ -31,13 +31,20 @@ When the verification must be performed, I<callback> will be called with
 the arguments callback(X509_STORE_CTX *x509_store_ctx, void *arg). The
 argument I<arg> is specified by the application when setting I<callback>.
 
-I<callback> should return 1 to indicate verification success and 0 to
-indicate verification failure. If SSL_VERIFY_PEER is set and I<callback>
-returns 0, the handshake will fail. As the verification procedure may
-allow the connection to continue in the case of failure (by always
-returning 1) the verification result must be set in any case using the
-B<error> member of I<x509_store_ctx> so that the calling application
-will be informed about the detailed result of the verification procedure!
+I<callback> should return 1 to indicate verification success
+and 0 to indicate verification failure.
+If B<SSL_VERIFY_PEER> is set and I<callback> returns 0, the handshake will fail.
+The I<callback> may also return -1 to indicate that the handshake
+shall suspend and return control to the calling application
+with B<SSL_ERROR_WANT_RETRY_VERIFY>.
```

Comment:
AFAICT, this part of the PR documentation is lying: The handshake will _not_ be suspended unless SSL_VERIFY_PEER[1] is set. It is very important to document this hidden fact IMO: The validation is performed unconditionally, but its `-1` (i.e. suspension) result is treated as a `1` (i.e. success) result unless SSL_VERIFY_PEER[1] is set!

I admit that the original documentation did not explicitly say that the validation failure will be ignored when the callback returns `0` without SSL_VERIFY_PEER[1]. However, I do not think we should make matters worse by hiding even more details about the special no-SSL_VERIFY_PEER[1] case.

Please note that I am not sure that we should special-case SSL_VERIFY_PEER when suspending. In fact, I feel like we should not, but this is not my area of expertise. If we should not suspend, then the PR _code_ needs to be changed (and this documentation may need to be updated differently to avoid questions about SSL_VERIFY_PEER effect on `-1`). I left another change request to remind you to check this logic.

---
[1] To be more precise, per the corresponding "The documented interface is that SSL_VERIFY_PEER..." source code comment, setting other flags is sufficient, but let's keep _that_ Pandora box closed in this PR!

Reply:
Goodness - at this point the existing documentation was pretty buggy.
I've cleaned this up except for the not really correct phrase "`if B<SSL_VERIFY_PEER> is set`",
since according to the code comment you mention an earlier cleanup attempt failed.

Still I've corrected other aspects of the existing description:

- Make clear the callback is called whenever a peer certificate has been received,
      which is independent of the verification mode.
- Make clear that a return value > 1 always leads to handshake failure.
- Make clear that in server mode also return values < 0 lead to handshake failure.
- Refer to X509_STORE_CTX_set_error() rather than to internal error variable (as already mentioned above).

Moreover, I've corrected my documentation of the new feature on `-1` 
in the sense that it is only relevant in client mode and "`if B<SSL_VERIFY_PEER> is set`".

Reply:
On second thought, 
I've moved the corrections of `SSL_CTX_set_cert_verify_callback.pod` to a different PR: #13937 
and made the new feature independent of the verification mode.

Reply:
Good call on moving unrelated documentation improvements into a separate PR! Judging by the initial feedback there, it may be difficult to untangle that documentation mess. Hopefully, we can keep the two PRs separate.